File: occasum-core/src/main/java/pw/stamina/occasum/properties/collection/AbstractCollectionProperty.java
package pw.stamina.occasum.properties.collection;

import pw.stamina.occasum.properties.AbstractProperty;
import pw.stamina.occasum.properties.PropertyParseException;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Objects;

public abstract class AbstractCollectionProperty<T, R extends Collection<T>>
        extends AbstractProperty implements CollectionProperty<T> {
    protected final R elements;
    private final Collection<T> defaultElements;
    private final CollectionPropertyAdapter<T> adapter;

    private String valueAsString;
    private String defaultValueAsString;

    protected AbstractCollectionProperty(String name, R elements,
                                         CollectionPropertyAdapter<T> adapter) {
        super(name);

        Objects.requireNonNull(elements, "elements");
        Objects.requireNonNull(adapter, "adapter");

        this.elements = elements;
        this.adapter = adapter;

        this.defaultElements = createDefaultElements(elements);
    }

    @Override
    public boolean add(T element) {
        Objects.requireNonNull(element, "element");
        invalidateValueAsString();
        return elements.add(element);
    }

    @Override
    public boolean remove(T element) {
        Objects.requireNonNull(element, "element");
        invalidateValueAsString();
        return elements.remove(element);
    }

    @Override
    public boolean contains(T element) {
        Objects.requireNonNull(element, "element");
        return elements.contains(element);
    }

    @Override
    public void parseAndSet(String input) throws PropertyParseException {
        Collection<T> parsed = adapter.parse(input);
        clearAndAddAll(parsed);
    }

    @Override
    public boolean isDefault() {
        return !elements.isEmpty();
    }

    @Override
    public void reset() {
        clearAndAddAll(defaultElements);
    }

    private void clearAndAddAll(Collection<T> collection) {
        elements.clear();
        elements.addAll(collection);
    }

    @Override
    public String getValueAsString() {
        if (valueAsString == null) {
            valueAsString = adapter.toString(elements);
        }

        return valueAsString;
    }

    @Override
    public String getDefaultValueAsString() {
        if (defaultValueAsString == null) {
            defaultValueAsString = adapter.toString(elements);
        }

        return defaultValueAsString;
    }

    private void invalidateValueAsString() {
        valueAsString = null;
    }

    private static <T> Collection<T> createDefaultElements(Collection<T> elements) {
        if (elements.isEmpty()) {
            return Collections.emptySet();
        } else {
            return new ArrayList<>(elements);
        }
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/collection/AbstractCollectionProperty.kt
package pw.stamina.occasum.properties.collection

import pw.stamina.occasum.properties.AbstractProperty
import pw.stamina.occasum.properties.PropertyParseException

import java.util.ArrayList
import java.util.Collections
import java.util.Objects

abstract class AbstractCollectionProperty<T, R : MutableCollection<T>> protected constructor(
        name: String,
        override val elements: R,
        private val adapter: CollectionPropertyAdapter<T>)
    : AbstractProperty(name), CollectionProperty<T> {

    private val defaultElements: Collection<T>

    private var valueAsString: String? = null
    private var defaultValueAsString: String? = null

    override val isDefault: Boolean
        get() = !elements.isEmpty()

    init {
        this.defaultElements = createDefaultElements(elements)
    }

    override fun add(element: T): Boolean {
        invalidateValueAsString()
        return elements.add(element)
    }

    override fun remove(element: T): Boolean {
        invalidateValueAsString()
        return elements.remove(element)
    }

    override fun contains(element: T): Boolean {
        return elements.contains(element)
    }

    @Throws(PropertyParseException::class)
    override fun parseAndSet(input: String) {
        val parsed = adapter.parse(input)
        clearAndAddAll(parsed)
    }

    override fun reset() {
        clearAndAddAll(defaultElements)
    }

    private fun clearAndAddAll(collection: Collection<T>) {
        elements.clear()
        elements.addAll(collection)
    }

    override fun getValueAsString(): String {
        if (valueAsString == null) {
            valueAsString = adapter.toString(elements)
        }

        return valueAsString
    }

    override fun getDefaultValueAsString(): String {
        if (defaultValueAsString == null) {
            defaultValueAsString = adapter.toString(elements)
        }

        return defaultValueAsString
    }

    private fun invalidateValueAsString() {
        valueAsString = null
    }

    private fun <T> createDefaultElements(elements: Collection<T>): Collection<T> {
        return if (elements.isEmpty()) {
            emptySet()
        } else {
            ArrayList(elements)
        }
    }
}
